Doc/c-api/exceptions.rst

```diff
@@ -868,13 +879,18 @@ The following functions are used to create and modify Unicode exceptions from C.
    *\*end*.  *end* must not be ``NULL``.  Return ``0`` on success, ``-1`` on
    failure.
 
+   If the :attr:`UnicodeError.object` is an empty sequence, the resulting
+   *end* is ``0``. Otherwise, it is clipped to ``[1, len(object)]``.
```

Comment:
Is there a particular reason why 0 isn't allowed? Empty substrings (``start==end``) can be useful for special cases.
OTOH, this PR allows “negative” substrings (``start>end``), which does *not* sound useful -- and it can throw off code like `length = start-end`.

IMO, `end` should be clipped to ``[clipped(start), len(object)]``.

Reply:
> Is there a particular reason why 0 isn't allowed

The original code did:

```C
    size = PyUnicode_GET_LENGTH(obj);
    if (*end<1)
        *end = 1;
    if (*end>size)
        *end = size;
    Py_DECREF(obj);
    return 0;
```

So it always clipped it to `[1, len(object)]` but put it at 0 for empty sequences. I just documented that behaviour correctly (otherwise it could be confusing). 

> OTOH, this PR allows “negative” substrings (start>end),

I don't know whether I should enforce it or not (we didn't decide on this matter, I only kept what was already there). 

> IMO, end should be clipped to [clipped(start), len(object)].

This could be a good alternative. I haven't thought of it, so I'll leave this comment opened to discuss your proposal.

Reply:
Looks like not much discussion is coming.
Is there anyone whose opinion you'd like to hear?

Reply:
Personally, I'd like to know what Victor and Serhiy think on that. The problem is I'm worried to break things even though we are fixing a bug. However, whatever we choose should not make the following crash:

```
str(UnicodeEncodeError('utf-8', '', -1, 0, ''))
```

Maybe we can start by fixing `__str__`? what do you think? (without touching on the getter/setters for now?). I once fixed it by changing the getter/setter implementation but then we didn't really know how to proceed forward.

Reply:
A straightforward fix for `__str__` sounds like it'd be easy to both review and backport.

Reply:
I'll work on that then.

Reply:
See https://github.com/python/cpython/pull/124935